### PR TITLE
Bigtable: Introducing new Batching API for Bigtable

### PR DESCRIPTION
**Pull request by rahulKQL**
Working towards #3003 

- Introducing new API `BigtableDataClient#newBulkMutationBatcher(String)`.
 - Deprecated existing API `BigtableDataClient#newBulkMutationBatcher()`.
 - Updated BatchingCallSettings to be compatible with new Batching API.
 - Removed description from deprecated methods/classes.
 - Added a new model in RowMutationEntry.
 - Added MutateRowsBatchingDescriptorV2 to spool the batching response.
 - Updated JavaDocs, test cases according to latest changes.

### CODE REVIEW

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/BulkMutation.java [hunk 1]
```diff
@@ -100,4 +110,12 @@ public MutateRowsRequest toProto(RequestContext requestContext) {
         .setAppProfileId(requestContext.getAppProfileId())
         .build();
   }
+
+  /** Creates a copy of {@link BulkMutation}. */
+  @Override
+  public BulkMutation clone() {
+    BulkMutation bulkMutation = BulkMutation.create(tableId);
+    bulkMutation.builder = this.builder.build().toBuilder();
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/BulkMutation.java]: proto builders natively support clone:
https://github.com/googleapis/google-cloud-java/blob/master/google-api-grpc/proto-google-cloud-bigtable-v2/src/main/java/com/google/bigtable/v2/MutateRowsRequest.java#L651-L653

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/RowMutationEntry.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+ * <p>Note: The changes in the mutation will be applied atomically and the ordering between rows is
+ * not guaranteed.
+ */
+public class RowMutationEntry implements MutationApi<RowMutationEntry>, Serializable {
+  private static final long serialVersionUID = 1974738836742298192L;
+
+  private final ByteString key;
+  private final Mutation mutation;
+
+  private RowMutationEntry(ByteString key, Mutation mutation) {
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/RowMutationEntry.java]: Please add preconditions & annotations for Nonnull

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java [hunk 3]
```diff
@@ -362,6 +366,34 @@ public static EnhancedBigtableStub create(EnhancedBigtableStubSettings settings)
...
+   *   <li>Process the response and schedule retries. At the end of each attempt, entries that have
+   *       been applied, are filtered from the next attempt. Also, any entries that failed with a
+   *       nontransient error, are filtered from the next attempt. This will continue until there
+   *       are no more entries or there are no more retry attempts left.
+   *   <li>Wrap batch failures in a {@link
+   *       com.google.cloud.bigtable.data.v2.models.MutateRowsException}.
+   *   <li>Split the responses using {@link MutateRowsBatchingDescriptorV2}.
+   * </ul>
+   */
+  public Batcher<RowMutationEntry, Void> newMutateRowsBatcher(String tableId) {
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java]: `@Nonnull`

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/RowMutationEntryBatcherIT.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    for (int i = 0; i < 10; i++) {
+      expectedRows.add(
+          Row.create(
+              ByteString.copyFromUtf8(rowPrefix + "-" + i),
+              Lists.newArrayList(
+                  RowCell.create(
+                      family,
+                      ByteString.copyFromUtf8("qualifier"),
+                      10_000L,
+                      Lists.<String>newArrayList(),
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/RowMutationEntryBatcherIT.java]: `ImmutableList.of()`

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java [hunk 5]
```diff
@@ -131,7 +129,8 @@
   private final UnaryCallSettings<Query, Row> readRowSettings;
   private final UnaryCallSettings<String, List<KeyOffset>> sampleRowKeysSettings;
   private final UnaryCallSettings<RowMutation, Void> mutateRowSettings;
-  private final BatchingCallSettings<RowMutation, Void> bulkMutateRowsSettings;
+  private final BatchingCallSettings<RowMutationEntry, Void, BulkMutation, Void>
+      bulkMutateRowsSettings;
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java]: As discussed offline, this is pending investigation. We need to future proof this for adding latency based throttling. Before we submit this PR, we should have a plan for configuring the throttle latency

**Comment by igorbernstein2** [comment]
Can you investigate why the linkage monitor is complaining?

**Comment by rahulKQL** [comment]
It seems the linkage monitor build was not able to find new `BatchingDescirptor`, Although the build was successful. I will update more information here after checking again.

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java (hunk 5 above)
Comment (rahulKQL) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java]: I am on it, Once we finalize on approach for this we can merge this.

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/RowMutationEntry.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+ */
+public class RowMutationEntry implements MutationApi<RowMutationEntry>, Serializable {
+  private static final long serialVersionUID = 1974738836742298192L;
+
+  private final ByteString key;
+  private final Mutation mutation;
+
+  private RowMutationEntry(@Nonnull ByteString key, @Nonnull Mutation mutation) {
+    Preconditions.checkNotNull(key);
+    Preconditions.checkNotNull(mutation);
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/RowMutationEntry.java]: Please add messages

**Comment by rahulKQL** [comment]
@igorbernstein2 @kolea2 I have added BigtableBatchingCallSettings, Please have a look and let me know your thoughts on this.

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/RowMutationEntry.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+  private RowMutationEntry(@Nonnull ByteString key, @Nonnull Mutation mutation) {
+    Preconditions.checkNotNull(key, "Row key cannot be null");
+    Preconditions.checkNotNull(mutation, "Row mutation cannot be null");
+
+    this.key = key;
+    this.mutation = mutation;
+  }
+
+  /** Creates a new instance of the mutation builder. */
+  public static RowMutationEntry create(@Nonnull String key) {
```
[the remark this answers is not in the document]
Comment (rahulKQL) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/RowMutationEntry.java]: I think that would be a little verbose considering there are many overloaded method. Also, we follow similar approach across existing models in `google-cloud-bigtable`.

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/BigtableBatchingCallSettings.java [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+import com.google.api.gax.rpc.StatusCode;
+import com.google.api.gax.rpc.UnaryCallSettings;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import java.util.Set;
+
+/**
+ * This settings is just a simple wrapper for {@link
+ * com.google.api.gax.batching.BatchingCallSettings} to allow us to add additional functionality.
+ * This settings holds the batching thresholds as well as retry configuration.
```
[the remark this answers is not in the document]
Comment (rahulKQL) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/BigtableBatchingCallSettings.java]: Could you please describe me what do you mean by code comment?

**Comment by rahulKQL** [comment]
The linkage build is failing for below classes which introduced in `gax-1.48.0`, and the `libraries-bom-2.2.1` points to `gax-1.44.0`.

```sh
BOM Coordinates: com.google.cloud:libraries-bom:2.2.1
Newly introduced problems:
Class com.google.api.gax.batching.BatchingDescriptor is not found
  referenced from com.google.cloud.bigtable.data.v2.stub.BigtableBatchingCallSettings (google-cloud-bigtable-0.106.1-SNAPSHOT.jar)
  referenced from com.google.cloud.bigtable.data.v2.stub.mutaterows.MutateRowsBatchingDescriptorV2 (google-cloud-bigtable-0.106.1-SNAPSHOT.jar)
Class com.google.api.gax.batching.BatchingRequestBuilder is not found
  referenced from com.google.cloud.bigtable.data.v2.stub.mutaterows.MutateRowsBatchingDescriptorV2 (google-cloud-bigtable-0.106.1-SNAPSHOT.jar)
Class com.google.api.gax.batching.BatcherImpl is not found
  referenced from com.google.cloud.bigtable.data.v2.stub.EnhancedBigtableStub (google-cloud-bigtable-0.106.1-SNAPSHOT.jar)
Class com.google.api.gax.batching.BatchingCallSettings is not found
  referenced from com.google.cloud.bigtable.data.v2.stub.BigtableBatchingCallSettings (google-cloud-bigtable-0.106.1-SNAPSHOT.jar)
Class com.google.api.gax.batching.BatchingCallSettings$Builder is not found
  referenced from com.google.cloud.bigtable.data.v2.stub.BigtableBatchingCallSettings (google-cloud-bigtable-0.106.1-SNAPSHOT.jar)
Exception in thread "main" com.google.cloud.tools.dependencies.linkagemonitor.LinkageMonitorException: Found 5 new linkage errors
	at com.google.cloud.tools.dependencies.linkagemonitor.LinkageMonitor.run(LinkageMonitor.java:105)
	at com.google.cloud.tools.dependencies.linkagemonitor.LinkageMonitor.main(LinkageMonitor.java:63)
```

**Pull request merged**